[Spanner] Adds support for retrying Spanner transactions on commit abort.

amanda-tarafa opened the pull request:
@jskeet just a quick look for now, sanity checking and that everything is where it should be. Taking a look a commit at a time might be helpfull.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerException.cs:
```diff
@@ -84,6 +86,25 @@ public bool IsRetryable
+        /// <summary>
+        /// For retryable <see cref="SpannerException"/>, this value represents
+        /// the recommended minimum retry delay. It might or might not
+        /// be present, even if this exeption is retryable.
```
Nit in case I miss it later: exeption => exception

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerException.cs:
```diff
@@ -84,6 +86,25 @@ public bool IsRetryable
+
+        /// <summary>
+        /// For retrying aborted commits, we need to diferentiate whether the operation
+        /// was aborted due to a session expiry. If the session expired we cannot
```
Can we not retry, or can we just not retry with the same session? I thought it was just more efficient if we reuse the sesssion.

jskeet replied:
Okay, I think I agree :)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerException.cs:
```diff
@@ -101,7 +122,20 @@ internal SpannerException(ErrorCode code, RpcException innerException)
+            if (retryInfoEntry != null)
+            {
+                RetryInfo retryInfo = RetryInfo.Parser.ParseFrom(retryInfoEntry.ValueBytes);
+                if (retryInfo?.RetryDelay?.Nanos > 0)
```
Why do we not want to use a retry delay of 0? (This may well be moving code I wrote before, but we should check :)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerRetryHelpers.cs:
```diff
@@ -0,0 +1,75 @@
+                }
+                catch (SpannerException e) when (retryOptions.ExceptionFilter(e))
+                {
+                    // If there's a recommended retry dely specified on the exception
```
dely => delay

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs:
```diff
@@ -143,6 +199,26 @@ internal static PooledSession FromSessionName(SessionPool.ISessionPool pool, Ses
+        /// Usually disposing of a <see cref="PooledSession"/> instance should release its
+        /// underlying session to the pool. But when we create a new instance of <see cref="PooledSession"/>
+        /// from an existing instance of <see cref="PooledSession"/>, that is, with the same underlying session,
+        /// the original instance shouldn't be allowed to be disposed off normally so that the underlying session
```
Nit: off => of

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs:
```diff
@@ -412,8 +415,37 @@ private async Task<PooledSession> BeginTransactionAsync(PooledSession session, T
+                    }
+                    catch(RpcException e)
+                    {
+                        Parent._logger.Warn("Failed to create transaction for acquiered session", e);
```
Nit: acquiered => acquired

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs:
```diff
@@ -412,8 +415,37 @@ private async Task<PooledSession> BeginTransactionAsync(PooledSession session, T
+                    {
+                        Parent._logger.Warn("Failed to create transaction for acquiered session", e);
+                        // Failed to create the transaction for the session.
+                        // Release the session back to the pool.
```
Possibly add that we'll have another go with another session?

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs:
```diff
@@ -412,8 +415,37 @@ private async Task<PooledSession> BeginTransactionAsync(PooledSession session, T
+                        session.ReleaseToPool(false);
+                    }
                 }
+                // If we are here we need to acquiere a new session.
```
Nit: acquiere => acquire

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -90,6 +90,21 @@ public override string ConnectionString
         [Category("Data")]
         public string SpannerInstance => Builder.SpannerInstance;
 
+        /// <summary>
```
We should probably ask the Spanner team about this. There's too much I don't know.

If we expose it at all though, I'd be tempted to make it a parameter on the method instead of having it as a property. (With a suitable default if null is passed in.)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -261,6 +276,76 @@ public SpannerTransaction BeginReadOnlyTransaction(TransactionId transactionId)
+        /// to be called more than once.
+        /// A new <see cref="SpannerTransaction"/> will be passed to <paramref name="asyncWork"/>
+        /// each time it is rerun.
+        /// <paramref name="asyncWork"/> doens't need to handle the lifecycle of the <see cref="SpannerTransaction"/>,
```
Nit: doens't => doesn't

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -261,6 +276,76 @@ public SpannerTransaction BeginReadOnlyTransaction(TransactionId transactionId)
+                        {
+                            if (transaction != null)
+                            {
+                                // Let's make sure that the associated session is not released to the pool
```
Why? I may be missing something.

amanda-tarafa commented:
Thannks! Will address comments and questions, and get ready for merge. Will flag when I'm done for a full review.

amanda-tarafa commented:
Travis failed because the compatibility checker failed to install dotnet. Rerunning.

amanda-tarafa commented:
This is now ready for review. A few notes:

- Reviewing one commit at a time might be helpful.
- I din't update the integration tests in DMLTests.cs to use the new RunWithRetriableTransaction... method. I created a new RetriableTransactionTests.cs instead. I figured that even though the recommended way to do writes is with a retriable transaction, client code can still use implicit transactions or manage their own transactions so it is worth it to keep those tests as well.
- I did update all the snippets that were writing data so that they'd use the RunWithRetriableTransaction... methods. Those are the docs, and we should show the recommended way of using read-write transactions.
- I removed the snippets having to do with System.Transactions.TransactionScope since that's not compatible with the recommended RunWithRetriableTransaction... The integrations tests for TransactionScope are unchanged.

amanda-tarafa commented:
Checking CI now.

amanda-tarafa commented:
Checking CI now.

amanda-tarafa commented:
CI is now green.

jskeet commented:
Okay, I'm re-reviewing from scratch, one commit at a time :)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerException.cs:
```diff
@@ -84,6 +86,25 @@ public bool IsRetryable
+        /// For retrying aborted commits, we need to diferentiate whether the operation
+        /// was aborted due to a session expiry. If the session expired we cannot
+        /// retry automatically because we cannot reuse the session.
+        /// Note that it is us who are setting the <see cref="ErrorCode"/> to <see cref="ErrorCode.Aborted"/>
```
I think this part would be better as a non-XML comment, just in case this property is ever made public.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerRetryHelpers.cs:
```diff
@@ -0,0 +1,109 @@
+            }
+        }
+
+        private class MinimumBoundJitter : IJitter
```
I don't think it's a good idea to implement `IJitter` here. It doesn't fulfill the same contract, as is evident by the fact that the parameter is `minDelay` whereas in `IJitter` it's `maxDelay`. I think just a simple method (passing in `retryOptions.DelayJitterMultiplier`) would be better. Apologies for having missed this before.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerRetryHelpers.cs:
```diff
@@ -0,0 +1,109 @@
+    /// TODO(atarafamas, jskeet): Consider making public or entirely eliminating once we know 
+    /// what we can reuse from GAX retry helpers.
+    /// </summary>
+    internal static class SpannerRetryHelpers
```
Just as a heads-up, we'll be changing some of the retry code in GAX soon. Maybe we want to bear this (and some Firestore code) in mind when doing that, to make it more amenable to reuse.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerRetryHelpers.cs:
```diff
@@ -0,0 +1,109 @@
+                {
+                    // If there's a recommended retry delay specified on the exception
+                    // we should at least wait for that.
+                    if (e.RecommendedRetryDelay.HasValue && e.RecommendedRetryDelay.Value > retryDelay)
```
It's not clear to me whether we need any JITter on top of that at all.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerRetryHelpers.cs:
```diff
@@ -0,0 +1,109 @@
+                    jitterMultiplier = s_random.NextDouble();
+                }
+
+                var minDelayTicks = minDelay.Ticks;
```
I don't think I fully understand what this code is meant to do, but I suspect it can be simplified. It'll be clearer with more comments about the purpose of the class/method.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerRetryOptions.cs:
```diff
@@ -0,0 +1,46 @@
+    /// </summary>
+    internal sealed class SpannerRetryOptions
+    {
+        internal SpannerRetryOptions(
```
Do we need all of this anyway? For example, do we ever really need to change the exception filter?

amanda-tarafa replied:
For retrying aborted commints:

- We shouldn't ever change the exception filter.
- As per discussion with the Spanner team, the backoff config is not modifiable by client code either.

But, I was trying to write a more general use kind of retry helper in case there are more "high level" retries we needed/wanted to offer, something that in the future, if needed across libraries, we could consider moving to Gax making the appropiate changes.

That said, I have no problem hardwiring some or all of this, but then I'd also like to have the retry method be specifically for aborted transactions and not for anything else. It would be weird to have the general method use the hardwire config (for instance, for the filter) that only really makes sense for aborted transactions.

Or should we consider modifying Gax to extract general retry logic that can be reused on a per call retry and also on higher level retries?

jskeet replied:
My experience is that there tends to be enough that's "just a bit different" in each application-level retry that it's not worth trying to encapsulate too much of it.
If there are aspects of GAX that can be easily exposed, I'm open to that.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerRetryHelpersTests.cs:
```diff
@@ -0,0 +1,249 @@
+
+namespace Google.Cloud.Spanner.Data.Tests
+{
+    public class SpannerRetryHelpersTests
```
Note: I haven't looked through the tests carefully, partly because I suspect they'll change based on other comments.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.TargetedSessionPoolTests.cs:
```diff
@@ -521,6 +509,80 @@ public async Task AcquireAsync_NursesBackToHealth_OriginalCallerCancelled()
+                await client.Scheduler.RunAsync(async () =>
+                {
+                    var originalSession = await pool.AcquireSessionAsync(
+                        new TransactionOptions (),
```
Nit: remove space before `()`

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.TargetedSessionPoolTests.cs:
```diff
@@ -822,13 +884,18 @@ public async Task ShutdownPoolAsync_CannotAcquireAfterShutdown()
                 return tasks.Select(t => t.Result).ToList();
             }
+
+            private List<Task<PooledSession>> AcquireSessionTasks(TargetedSessionPool pool, int sessionsToAcquiere)
```
Optional nit: use expression-bodied member?

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs:
```diff
@@ -58,6 +58,24 @@ public sealed class PooledSession : IDisposable
         /// </summary>
         internal ModeOneofCase TransactionMode { get; }
 
+        internal TransactionOptions TransactionOptions
```
Optional nit: might want to make this a method, given that it creates a new instance each time.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs:
```diff
@@ -27,6 +27,7 @@ internal interface ISessionPool
         {
             SpannerClient Client { get; }
             IClock Clock { get; }
+            Task<PooledSession> WithFreshTransactionOrNewAsync(PooledSession session, TransactionOptions transactionOptions, CancellationToken cancellationToken);
```
I'm glad ISessionPool is internal, otherwise this would have been a breaking change :)

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs:
```diff
@@ -427,11 +427,14 @@ private async Task TryCreateReadWriteTransactionAndReturnToPool(PooledSession se
             }
 
-            private async Task<PooledSession> BeginTransactionAsync(PooledSession session, TransactionOptions options, CancellationToken cancellationToken)
+            private async Task<PooledSession> BeginTransactionAsync(PooledSession session, TransactionOptions options, CancellationToken cancellationToken, bool isSessionAcquired = false)
```
I don't fully understand what this parameter is yet. It's a little odd that nothing else has changed in the method, apart from whether or not the book-keeping. Maybe add a comment explaining this more?

jskeet replied:
Yes, better thanks.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs:
```diff
@@ -443,8 +446,38 @@ private async Task<PooledSession> BeginTransactionAsync(PooledSession session, T
+
+            public override Task<PooledSession> WithFreshTransactionOrNewAsync(PooledSession session, TransactionOptions transactionOptions, CancellationToken cancellationToken)
+            {
+                if(session.RequiresRefresh)
```
Nit: space after `if`.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -261,6 +274,104 @@ public SpannerTransaction BeginReadOnlyTransaction(TransactionId transactionId)
             BeginTransactionImplAsync(s_readWriteTransactionOptions, TransactionMode.ReadWrite, cancellationToken);
 
+        /// <summary>
+        /// Executes a read-write transaction, with retries as necessary.
```
Possibly clarify in "remarks" what's meant by "as necessary"? (It's only retrying certain things.)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -261,6 +274,104 @@ public SpannerTransaction BeginReadOnlyTransaction(TransactionId transactionId)
+        public async Task RunWithRetriableTransactionAsync(Func<SpannerTransaction, Task> asyncWork, CancellationToken cancellationToken = default)
+        {
+            GaxPreconditions.CheckNotNull(asyncWork, nameof(asyncWork));
+            await RunWithRetriableTransactionAsync(async (transaction) =>
```
Nit on style: remove parentheses from `(transaction)`.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -261,6 +274,104 @@ public SpannerTransaction BeginReadOnlyTransaction(TransactionId transactionId)
+                session?.Dispose();
+            }
+
+            async Task<TResult> CommitAttempt()
```
It feels like this could be in SpannerRetryHelpers. It's possible there's too much context here to easily move it there, but it would be nice to make the retry helper class be more helpful :) That can definitely happen in a later PR though - maybe add a TODO for now?

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -261,6 +274,104 @@ public SpannerTransaction BeginReadOnlyTransaction(TransactionId transactionId)
+            try
+            {
+                await OpenAsync(cancellationToken).ConfigureAwait(false);
+                return await SpannerRetryHelpers.ExecuteWithRetryAsync(
```
I think it would be worth noting in here that `CommitAttempt` modifies the `session` variable, otherwise this looks a bit odd - it looks like session will always be null.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -99,17 +97,20 @@ public async Task InsertDataAsync()
 
+                // If the transaction is aborted, RunWithRetriableTransactionAsync will
+                // retry the whole unit of work with a fresh transaction each time.
+                await connection.RunWithRetriableTransactionAsync(async (transaction) =>
```
Nit: remove parentheses from `(transaction)`

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -99,17 +97,20 @@ public async Task InsertDataAsync()
                     SpannerParameter keyParameter = cmd.Parameters.Add("Key", SpannerDbType.String);
                     SpannerParameter stringValueParameter = cmd.Parameters.Add("StringValue", SpannerDbType.String);
                     SpannerParameter int64ValueParameter = cmd.Parameters.Add("Int64Value", SpannerDbType.Int64);
+                    cmd.Transaction = transaction;
```
The comment below this line is now incorrect, isn't it?

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -294,61 +315,46 @@ public async Task ReadUpdateDeleteAsync()
+                        });
+                    cmd.Transaction = transaction;
+
+                    // This executes a single transactions with alls row written at once during CommitAsync().
```
alls row => all rows

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -294,61 +315,46 @@ public async Task ReadUpdateDeleteAsync()
-            // Additional: BeginTransactionAsync
-            RetryPolicy<SpannerFaultDetectionStrategy> retryPolicy =
-                new RetryPolicy<SpannerFaultDetectionStrategy>(RetryStrategy.DefaultExponential);
+            // Sample: RetriableTransactionAsync
```
Is this now just the same as InsertDataAsync?

jskeet commented on apis/Google.Cloud.Spanner.Data/docs/index.md:
```diff
@@ -128,29 +146,20 @@ Cloud Spanner is fully ACID compliant.
 
 {{sample:SpannerConnection.TransactionAsync}}
 
-## TransactionScope support (.NET 4.5+, Single resource only)
```
Let's talk offline about this being removed. I think it's reasonable to say we still support TransactionScope, but that it won't retry transactions automatically. If we don't want to support it at all any more, maybe we should think about removing it for the upcoming breaking change.

jskeet replied:
Let's make sure we do that. I suspect we really do need to keep TransactionScope support - which is in netstandard2.0, so we don't need to say ".NET 4.5+" any more - but that we need to emphasize the retry aspect.

amanda-tarafa commented:
@jskeet this is ready for review.
The first and last commit are mostly nit and minor fixes.
The second commit contains most new changes, since we agreed on retry logic specific for aborted transactions.
PTAL.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs:
```diff
@@ -143,6 +181,26 @@ internal static PooledSession FromSessionName(SessionPool.ISessionPool pool, Ses
+            bool wasDisposed = Interlocked.Exchange(ref _disposed, 1) == 1;
+            if (!wasDisposed)
+            {
+                GC.SuppressFinalize(this);
```
I don't think we even have a finalizer now - this may have been left over from a time that we did. Not suggesting a change right now, but we may be able to simplify this method to a single expression.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs:
```diff
@@ -114,10 +114,48 @@ internal static PooledSession FromSessionName(SessionPool.ISessionPool pool, Ses
+        /// <returns>A new instance of <see cref="PooledSession"/>.</returns>
+        /// <exception cref="InvalidOperationException">If this <see cref="PooledSession.TransactionMode"/>
+        /// is <see cref="ModeOneofCase.None"/>.</exception>
+        /// <remarks>
```
Nit: I'd move the remarks to just below summary.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs:
```diff
@@ -295,6 +351,21 @@ internal Task<Transaction> BeginTransactionAsync(BeginTransactionRequest request
             return RecordSuccessAndExpiredSessions(Client.BeginTransactionAsync(request, callSettings));
         }
 
+        internal TransactionOptions GetTransactionOptions()
```
These won't always be complete options - for example, I think read-only transactions usually have a staleness bound. (There may be other things too.)
Not sure how much of a problem this is. We should probably talk about this. Also, could this method be private?

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -99,19 +97,24 @@ public async Task InsertDataAsync()
 
-                    // This executes 5 distinct transactions with one row written per transaction.
+                    // This executes 5 distinct insert commands using the retriable transaction.
+                    // The mutations will be effective once the transaction has commited successfully.
```
Nit: commited => committed

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -120,31 +123,36 @@ public async Task InsertDataAsync()
+                // retry the whole unit of work with a fresh transaction each time.
+                // Please be aware that the whole unit of work needs to be prepared
+                // to be called more than once.
+                await connection.RunWithRetriableTransactionAsync(async (transaction) =>
```
Nit: take transaction out of parentheses?

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -120,31 +123,36 @@ public async Task InsertDataAsync()
+                {
                     SpannerCommand cmd = connection.CreateDmlCommand(
-                        "UPDATE TestTable SET StringValue='Updated' WHERE Int64Value=@value");
+                    "UPDATE TestTable SET StringValue='Updated' WHERE Int64Value=@value");
```
It's worth reindenting this, I think.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -194,44 +206,49 @@ public async Task BatchDml()
+                // retry the whole unit of work with a fresh transaction each time.
+                // Please be aware that the whole unit of work needs to be prepared
+                // to be called more than once.
+                await connection.RunWithRetriableTransactionAsync(async (transaction) =>
```
Again, remove parentheses around the parameter.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -242,6 +259,7 @@ public async Task CommitTimestampAsync()
 
                     // Display the inserted values
                     SpannerCommand selectCmd = connection.CreateSelectCommand("SELECT * FROM UsersHistory");
+                    selectCmd.Transaction = transaction;
```
I don't know whether we really want to do the display part within the same transaction. It's not even clear to me whether the first two should be in the same transaction.
Aside from anything else, if they haven't been committed yet, I don't know what the commit timestamp would be... I suspect we should have two retriable bits, then execute the SELECT not in a transaction at all. That should show us two different commit timestamps.

jskeet commented on apis/Google.Cloud.Spanner.Data/docs/index.md:
```diff
@@ -77,8 +83,14 @@ visible to all subsequent statements in the batch.
 executed. A `SpannerBatchNonQueryException` will be thrown that contains both information about the error
 and the number of rows affected by each of the statements that executed before the failed one. If you are executing
-the batch DML command inside a transaction you can simply commit the transaction when `SpannerBatchNonQueryException`
+the batch DML command inside a non retriable transaction you can simply commit the transaction when `SpannerBatchNonQueryException`
```
Nit: hyphenate non-retriable

jskeet commented on apis/Google.Cloud.Spanner.Data/docs/index.md:
```diff
@@ -128,28 +146,19 @@ Cloud Spanner is fully ACID compliant.
-strategy for the TransientFaultApplication Block.
+- The callback doesn't need to handle the transaction lifecycle.
+- The callback needs to set the transaction received on all the commands it creates.
+- The callback might be called more than once so it needs to be prepared for it.
```
Possibly "prepared" is an unfortunate term, due to the idea of preparing statements etc. How about:

"The callback might be called more than once, so the code should be written bearing this in mind."

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs:
```diff
@@ -0,0 +1,337 @@
+            // Combining backoff and jittering for retries
+            // when the RpcException doesn't specify a wait time.
+            // InitialDelay = 250ms, BackoffMultiplier = 2, JitterMultiplier = 2
+            // Those are all hardwire, client code can't change them.
```
Nit: hardwire => hardwired or hard-coded perhaps.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs:
```diff
@@ -0,0 +1,337 @@
+            // when the RpcException doesn't specify a wait time.
+            // InitialDelay = 250ms, BackoffMultiplier = 2, JitterMultiplier = 2
+            // Those are all hardwire, client code can't change them.
+            private const int SecondCallLowBo = 250;
```
Possibly expand "Bo" to "Backoff"? It took me a little while to work out what was meant.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs:
```diff
@@ -0,0 +1,337 @@
+                    Assert.Equal(1, result);
+                });
+
+                callee.AssertCallTimesInrange(new Tuple<DateTime, DateTime>(time0, time0));
```
If you want to use ValueTuple rather than Tuple, I'm fine to add a dependency. I'm also happy to commit as it is, then I'll have a look at simplifying in another PR :)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs:
```diff
@@ -0,0 +1,337 @@
+
+                callee.AssertCallTimesInrange(
+                    new Tuple<DateTime, DateTime>(time0, time0),
+                    new Tuple<DateTime, DateTime>(time0.AddMilliseconds(SecondCallLowBo), time0.AddMilliseconds(SecondCallHighBo)),
```
If you created TimeSpan equivalents of the constants, you could just use `time0 + SecondCallLowBackoff` etc. Get rid of 18 calls to AddMilliseconds :)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs:
```diff
@@ -0,0 +1,337 @@
+                    Assert.Equal(3, result);
+                });
+
+                callee.AssertCallTimesInrange(
```
I wonder whether this might be refactored to AssertBackoffTimesInRange or somthing like that?
In particular, that could be stricter - because the third time will depend on the second time, if you see what I mean. (We can chat about this.)

jskeet replied:
All of this could be done in another PR though :)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs:
```diff
@@ -0,0 +1,337 @@
+                    var exception = await Assert.ThrowsAsync<SpannerException>(() => connection.RunWithRetriableTransactionAsync(transaction => callee.DatabaseWorkAsync(connection, transaction)));
+                    Assert.True(exception.IsRetryable && !exception.SessionExpired);
+                    Assert.Contains("Bang!", exception.InnerException.Message);
+
```
Nit: remove empty line.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs:
```diff
@@ -0,0 +1,337 @@
+
+                // The overall deadline is of 1 hour. The maximum backoff delay is of 32s.
+                // Because of jitter retries can stop at any time after 60mins - 64s, let's give it 2 minutes of range.
+                Assert.InRange(scheduler.Clock.GetCurrentDateTimeUtc(), time0.AddMinutes(58), time0.AddMinutes(60));
```
Can we check that there's been at least 110 (or whatever the theoretical minimum is) calls?

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs:
```diff
@@ -0,0 +1,337 @@
+
+                await scheduler.RunAsync(async () =>
+                {
+                    var exception = await Assert.ThrowsAsync<SpannerException>(() => connection.RunWithRetriableTransactionAsync(transaction => callee.DatabaseWorkAsync(connection, transaction)));
```
Nit: possibly wrap the line just after `ThrowsAsync<SpannerException>(`

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs:
```diff
@@ -0,0 +1,337 @@
+                    throw new InvalidOperationException("Bang!");
+                }
+
+                public void AssertCallTimesInrange(params Tuple<DateTime, DateTime>[] times)
```
Nit: capitalize the R of Range - but we may change how all of this works anyway :)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs:
```diff
@@ -0,0 +1,337 @@
+
+                public void AssertCallTimesInrange(params Tuple<DateTime, DateTime>[] times)
+                {
+                    for (int i = 0; i < times.Length; i++)
```
Should we check that `times` is equal to `_callTimes.Count`?

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SpannerClientHelpers.cs:
```diff
@@ -12,30 +12,225 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+using Google.Api.Gax.Grpc;
```
In order to speed up this review, I'm not reviewing this file right now.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransaction.cs:
```diff
@@ -0,0 +1,135 @@
+            _options = options ?? RetriableTransactionOptions.CreateDefault();
+        }
+
+        public async Task<TResult> RunAsync<TResult>(Func<SpannerTransaction, Task<TResult>> asyncWork, CancellationToken cancellationToken = default)
```
Maybe make the method internal, given that it's effectively internal anyway?

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransaction.cs:
```diff
@@ -0,0 +1,135 @@
+
+                            return result;
+                        }
+                        catch when (transaction != null)
```
It took me a little while to work out how transaction *could* be null. Maybe add a comment?

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransaction.cs:
```diff
@@ -0,0 +1,135 @@
+                            }
+                            catch (Exception e)
+                            {
+                                // If the commit actually aborted, calling Rollback will fail.
```
I don't suppose we can tell whether that's the case and avoid even trying the rollback?

amanda-tarafa replied:
No, we really can't tell if it was the commit that failed or not, or we could but with uglier code than this I think. I did changed the comment though, because rollback will fail if the commit was attempted and failed, but not only on aborted but other possible failures.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransaction.cs:
```diff
@@ -0,0 +1,135 @@
+                            catch (Exception e)
+                            {
+                                // If the commit actually aborted, calling Rollback will fail.
+                                // We don't want that or any other rollback exception to propogate,
```
Nit: propogate => propagate

amanda-tarafa replied:
Done

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransaction.cs:
```diff
@@ -0,0 +1,135 @@
+                            if (transaction != null)
+                            {
+                                // Let's make sure that the associated session is not released to the pool
+                                // because we'll be attempting to get a fresh transaction for this same session first.
```
A few concerns here:

- Don't we want to release the session if the transaction has succeeded? We'll still be executing this code.
- What about if the transaction failed, but we're not going to retry either due to deadlines or it not being an abort?

jskeet replied:
Ah, my guess is that both of these are handled by the final `session?.Dispose()` call - please add a comment to avoid silly people like me getting confused again :)

amanda-tarafa replied:
Done :)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransaction.cs:
```diff
@@ -0,0 +1,135 @@
+                {
+                    // If there's a recommended retry delay specified on the exception
+                    // we should respect it.
+                    if (e.RecommendedRetryDelay.HasValue)
```
How about

```csharp
retryDelay = e.RecommendedRetryDelay ?? retryDelay;
```

amanda-tarafa replied:
Yep!

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransaction.cs:
```diff
@@ -0,0 +1,135 @@
+                        throw;
+                    }
+                    await _scheduler.Delay(actualDelay, cancellationToken).ConfigureAwait(false);
+                    retryDelay = _options.NextDelay(retryDelay);
```
If we took a recommended delay, do we definitely want to double *that* rather than "the delay we'd have been using otherwise"?

amanda-tarafa replied:
Apparently, having a retry backoff configuration is just for being on the very safe side, because on aborted the server will always recommend a delay.
I'm fine either way, but although I couldn't swear on it right now, I think I did this because it's what Java was doing.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransactionOptions.cs:
```diff
@@ -0,0 +1,71 @@
+        private RetriableTransactionOptions(Func<TimeSpan, TimeSpan> jitter = null) =>
+            _jitter = jitter ?? MinimumBoundJitter;
+
+        public static RetriableTransactionOptions CreateDefault() => new RetriableTransactionOptions();
```
Make all of these internal for clarity?

amanda-tarafa replied:
Done

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransactionOptions.cs:
```diff
@@ -0,0 +1,71 @@
+    internal sealed class RetriableTransactionOptions
+    {
+        private static readonly BackoffSettings s_defaultBackoff = new BackoffSettings(
+                delay: TimeSpan.FromMilliseconds(250),
```
Nit: outdent by 4

amanda-tarafa replied:
Done

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -261,6 +260,103 @@ public SpannerTransaction BeginReadOnlyTransaction(TransactionId transactionId)
+        /// <param name="asyncWork">The work to perform in each transaction attempt.</param>
+        /// <param name="cancellationToken">An optional token for canceling the call.</param>
+        /// <returns>The value returned by <paramref name="asyncWork"/> if the transaction commits successfully.</returns>
+        /// <remarks><paramref name="asyncWork"/> will be fully retried whenever the <see cref="SpannerTransaction"/>
```
Nit: move remarks to just after summary (and in the method below.)

amanda-tarafa replied:
Done

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerException.cs:
```diff
@@ -101,7 +119,20 @@ internal SpannerException(ErrorCode code, RpcException innerException)
+            if (retryInfoEntry != null)
+            {
+                RetryInfo retryInfo = RetryInfo.Parser.ParseFrom(retryInfoEntry.ValueBytes);
+                if (retryInfo?.RetryDelay?.Nanos >= 0)
```
I suspect this was copied from some of my code somewhere, but I think it's slightly broken - if the server specifies a whole number of seconds, this won't pick that up.
I'd convert it to a TimeSpan unconditionally, and only set the RecommendedRetryDelay if that's not TimeSpan.Zero.

amanda-tarafa replied:
Done. I'll look the code that was doing this then, because it probably needs to be fixed as well.

amanda-tarafa commented:
All comments addressed on the last 3 commits (I will squash those into the first 3 later). I'm running tests locally. PTAL.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/RetriableTransactionOptionsTests.cs:
```diff
@@ -24,38 +24,44 @@ public class RetriableTransactionOptionsTests
+            var expectedMinDelayTicks = delayTicks;
+            var expectedMaxDelayTicks = delayTicks * 2;
+            var possibleDelayValues = expectedMaxDelayTicks - expectedMinDelayTicks;
+            var expectedMedianDelayTicks = expectedMinDelayTicks + possibleDelayValues / 2;
```
I think this would be simpler as:

```csharp
var expectedMedianDelayTicks = (expectedMinDelayTicks + expectedMaxDelayTicks) / 2;
```

amanda-tarafa replied:
Changed!

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/RetriableTransactionOptionsTests.cs:
```diff
@@ -24,38 +24,44 @@ public class RetriableTransactionOptionsTests
+            var expectedMedianDelayTicks = expectedMinDelayTicks + possibleDelayValues / 2;
 
-            var delays = Enumerable.Range(0, (int)TimeSpan.TicksPerMillisecond)
+            var delays = Enumerable.Range(0, possibleDelayValues)
```
This is still using a value which is fundamentally "a number of ticks" as "the number of iterations to perform" - they don't seem related to me, or at least, I don't understand why they would be.

It feels like the number of iterations could be any reasonably large number (even just 1000) - all we're checking is that at least one value is in the lower half and at least one value is in the upper half. I may be missing something though.

amanda-tarafa replied:
> This is still using a value which is fundamentally "a number of ticks" as "the number of iterations to perform" - they don't seem related to me, or at least, I don't understand why they would be.

But ticks is as granular as TimeSpan would go. So, when we jitter, the quantity of possible values we can ever get is always the number of ticks between the jitter min and max bounds.
In a perfect random distribution we would get each of the possible values if we had max - min iterations, and that's why, yes, this is fundamentally a number of ticks, but it represents the amount of different values so it's ok to use it as the number of iterations to be performed. What is true is that, if the random generator is good, this tests is really loose, and we could probably do with way less iterations, or with being way more strict on our checks.

I would leave it as is, but have no problem with putting just a number there, LMK.

jskeet replied:
I'd definitely prefer to have just a number. Even with a perfect random number generator, you wouldn't necessarily get each value exactly once - just like if you roll a fair die 6 times, you wouldn't always expect to get 1, 2, 3, 4, 5, 6.
(To put it another way - if that *were* the case, the last sample *wouldn't* be random, because you'd be able to predict the exact value - that would be weird!)

It might be reasonable to take, say, 10,000 samples and assert that between 4,000 and 6,000 of them should be in the lower half of the range - that would be a reasonable test that it's probably genuinely linear.

amanda-tarafa replied:
Yep, true, it would be a perfect distribution, but not random :).
Changing the test now.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs:
```diff
@@ -183,12 +181,17 @@ public async Task CommitAbortsAlways_RespectsOverallDeadline()
                 Assert.InRange(scheduler.Clock.GetCurrentDateTimeUtc(), time0.AddMinutes(58), time0.AddMinutes(60));
+                // The minimum calls that can be made (assuming maximum jitter always) in that time is 60 * 60 / 32 * 2 which is 56.25,
+                // plus 1 because the first one has no delay.
+                // The maximum number of calls tha can be made in 1 hour:
```
Nit: tha => that, and on the line below, aprox => approx

amanda-tarafa replied:
Done

amanda-tarafa commented:
> As discussed, we should make this only valid for ReadWrite transactions. (It would be okay for PartitionedDml as well, as that has no extra details.)
> We should probably think about what we'd do if TransactionOptions.ReadWrite ever gained any fields though... one simple option would be to make this method _accept_ a TransactionOptions, which we could check was of the same type as the existing one. It may be simpler to do that now rather than later.

I've changed the method to take the parameter and do the check. Added a test for that as well.

amanda-tarafa commented:
Rerunning Travis, it failed because it couldn't download something from Nuget.

amanda-tarafa commented:
Last rebase, will wait for green before merging.
